### PR TITLE
bft_error_handling: Added apollo cases

### PR DESCRIPTION
**Pull request by situ-s**
bft_error_handling: Added apollo cases

### CODE REVIEW

File: tests/apollo/test_skvbc_bft_error.py [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+    SUCCESS = 0
+    UNKNOWN = 1
+    INVALID_REQUEST = 2
+    NOT_READY = 3
+    TIMEOUT = 4
+    EXEC_DATA_TOO_LARGE = 5
+    EXEC_DATA_EMPTY = 6
+    CONFLICT_DETECTED = 7
+    OVERLOADED = 8
+    INTERNAL_ERROR = 9
```
Comment (yuliasherman) [inline tests/apollo/test_skvbc_bft_error.py:31]: This needs to be moved to bft_msgs.py

File: tests/apollo/util/skvbc.py [hunk 2]
```diff
@@ -119,15 +119,16 @@ def get_block_data_req(cls, block_id):
...
+        if data[0] != b'':
+            (reply, offset) = skvbc_messages.SKVBCReply.deserialize(data[0])
+            if isinstance(reply.reply, skvbc_messages.SKVBCWriteReply):
+                return WriteReply(reply.reply.success, reply.reply.latest_block)
+            elif isinstance(reply.reply, skvbc_messages.SKVBCReadReply):
+                return dict(reply.reply.reads)
+            elif isinstance(reply.reply, skvbc_messages.SKVBCGetLastBlockReply):
+                return reply.reply.latest_block
+            else:
+                raise BadReplyError
```
Comment (yuliasherman) [inline tests/apollo/util/skvbc.py:131]: You have unpack_reply in bft_msgs.py; incorporate your changes there.

File: tests/apollo/test_skvbc_bft_error.py [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+            ]
+
+class SkvbcBftErrorTest(unittest.TestCase):
+    @with_trio
+    @with_bft_network(start_replica_cmd)
+    async def test_results_from_replicas(self, bft_network):
+        """
+        Test that a replica succeeds to ask for missing info from the former window
+
+        1. Start all replicas but the primary
```
Comment (yuliasherman) [inline tests/apollo/test_skvbc_bft_error.py:58]: Why not the primary? Copy/paste issue?

File: bftengine/src/bftengine/RequestHandler.cpp [hunk 4]
```diff
@@ -32,8 +32,10 @@ void RequestHandler::execute(IRequestsHandler::ExecutionRequestsQueue& requests,
                              std::optional<Timestamp> timestamp,
                              const std::string& batchCid,
                              concordUtils::SpanWrapper& parent_span) {
+  LOG_INFO(GL, "SS-- In RequestHandler::execute");
   for (auto& req : requests) {
     if (req.flags & KEY_EXCHANGE_FLAG) {
+      LOG_INFO(KEY_EX_LOG, "BFT handler result" << req.outExecutionStatus);
```
[the remark this answers is not in the document]
Comment (situ-s) [reply bftengine/src/bftengine/RequestHandler.cpp]: Thanks Yulia

File: tests/simpleKVBC/TesterReplica/internalCommandsHandler.cpp [hunk 5]
```diff
@@ -122,8 +120,10 @@ void InternalCommandsHandler::execute(InternalCommandsHandler::ExecutionRequests
                                 verUpdates,
                                 merkleUpdates);
     }
-    if (res != OperationResult::SUCCESS) LOG_WARN(m_logger, "Command execution failed!");
-    req.outExecutionStatus = static_cast<uint32_t>(res);
+    if (res != OperationResult::SUCCESS) {
+      LOG_WARN(m_logger, "Command execution failed!");
+      req.outExecutionStatus = static_cast<uint32_t>(res);
```
Comment (yuliasherman) [inline tests/simpleKVBC/TesterReplica/internalCommandsHandler.cpp:125]: Why did you change this? req.outExecutionStatus should be set in any case.

File: bftengine/src/bftengine/ReplicaImp.cpp [hunk 6]
```diff
@@ -4539,14 +4541,16 @@ void ReplicaImp::executeReadOnlyRequest(concordUtils::SpanWrapper &parent_span,
       reply.setReplyLength(actualReplyLength);
       reply.setReplicaSpecificInfoLength(actualReplicaSpecificInfoLength);
       send(&reply, clientId);
+      return;
     } else {
       LOG_WARN(GL, "Received zero size response. " << KVLOG(clientId));
```
Comment (yuliasherman) [inline bftengine/src/bftengine/ReplicaImp.cpp:4546]: Reply buffer should be filled by an error here.

File: bftengine/src/bftengine/ReplicaImp.cpp [hunk 7]
```diff
@@ -5496,10 +5503,6 @@ void ReplicaImp::sendResponses(PrePrepareMsg *ppMsg, IRequestsHandler::Execution
         req.outReply = nullptr;
         clientsManager->removePendingForExecutionRequest(req.clientId, req.requestSequenceNum);
         continue;
-      } else {
-        LOG_WARN(CNSUS, "Received zero size response." << KVLOG(req.clientId, req.requestSequenceNum, ppMsg->getCid()));
-        strcpy(req.outReply, "Executed data is empty");
-        executionResult = static_cast<uint32_t>(bftEngine::OperationResult::EXEC_DATA_EMPTY);
       }
```
Comment (yuliasherman) [inline bftengine/src/bftengine/ReplicaImp.cpp:5506]: Why did you remove an empty data handling?

File: client/clientservice/src/request_service.cpp [hunk 8]
```diff
@@ -47,27 +47,35 @@ Status RequestServiceImpl::Send(ServerContext* context, const Request* proto_req
       LOG_INFO(logger_, "Send returned error");
```
Comment (yuliasherman) [inline client/clientservice/src/request_service.cpp:47]: Remove this log

File: client/clientservice/src/request_service.cpp [hunk 9]
```diff
@@ -47,27 +47,35 @@ Status RequestServiceImpl::Send(ServerContext* context, const Request* proto_req
...
       switch (std::get<uint32_t>(send_result)) {
         case (static_cast<uint32_t>(bftEngine::OperationResult::INVALID_REQUEST)):
+          LOG_INFO(logger_, "Returned INVALID_ARGUMENT");
```
Comment (yuliasherman) [inline client/clientservice/src/request_service.cpp:50]: Replace by "Request failed with INVALID_ARGUMENT error for cid=" << req_config.correlation_id
etc.

File: tests/apollo/test_skvbc_reply.py [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+        key = skvbc.random_key()
+        value = skvbc.random_value()
+        kv_pair = [(key, value)]
+
+        reply = await client.write_with_result(skvbc.write_req([], kv_pair, 0), result=bft_msgs.OperationResult.SUCCESS)
+        assert reply[1] == bft_msgs.OperationResult.SUCCESS, \
+                        f"Expected Reply={bft_msgs.OperationResult.SUCCESS}; actual={reply[1]}"
+
+        reply = await client.write_with_result(skvbc.write_req([], kv_pair, 0), result=bft_msgs.OperationResult.UNKNOWN)
+        assert reply[1] == bft_msgs.OperationResult.UNKNOWN, \
```
Comment (yuliasherman) [inline tests/apollo/test_skvbc_reply.py:62]: You should never receive UNKNOWN result. Please remove this.

File: tests/simpleKVBC/TesterReplica/internalCommandsHandler.cpp [hunk 11]
```diff
@@ -90,8 +90,11 @@ void InternalCommandsHandler::execute(InternalCommandsHandler::ExecutionRequests
   auto pre_execute = requests.back().flags & bftEngine::PRE_PROCESS_FLAG;
 
   for (auto &req : requests) {
-    if (req.outExecutionStatus != static_cast<uint32_t>(OperationResult::UNKNOWN))
-      continue;  // Request already executed (internal or pre-executed)
+    if (req.outExecutionStatus != static_cast<uint32_t>(OperationResult::UNKNOWN) &&
+        (requests.back().flags & HAS_PRE_PROCESSED_FLAG)) {
+      LOG_INFO(m_logger, "Requests already executed");
```
Comment (yuliasherman) [inline tests/simpleKVBC/TesterReplica/internalCommandsHandler.cpp:95]: Nit: Requests -> Request

File: util/pyclient/bft_client.py [hunk 12]
```diff
@@ -128,22 +128,31 @@ async def _recv_data(self, required_replies, dest_replicas, cancel_scope):
     def get_total_num_replicas(self):
         return len(self.replicas)
 
-    async def write(self, msg, seq_num=None, cid=None, pre_process=False, m_of_n_quorum=None,
-                    reconfiguration=False, corrupt_params={}, no_retries=False):
+    async def write_with_result(self, msg, seq_num=None, cid=None, pre_process=False, m_of_n_quorum=None,
+                    reconfiguration=False, corrupt_params={}, no_retries=False, result=0):
```
Comment (yuliasherman) [inline util/pyclient/bft_client.py:132]: The default result value should be 1 (UNKNOWN)

File: tests/simpleKVBC/TesterReplica/internalCommandsHandler.cpp [hunk 13]
```diff
@@ -539,7 +547,6 @@ OperationResult InternalCommandsHandler::executeReadOnlyCommand(uint32_t request
     const uint8_t *request_buffer_as_uint8 = reinterpret_cast<const uint8_t *>(request);
     deserialize(request_buffer_as_uint8, request_buffer_as_uint8 + requestSize, deserialized_request);
   } catch (const runtime_error &e) {
-    outReplySize = 0;
     LOG_ERROR(m_logger, "Failed to deserialize SKVBCRequest: " << e.what());
     return OperationResult::INTERNAL_ERROR;
```
Comment (yuliasherman) [inline tests/simpleKVBC/TesterReplica/internalCommandsHandler.cpp:551]: The output buffer should be filled here by an error

File: client/clientservice/src/request_service.cpp [hunk 14]
```diff
@@ -44,30 +44,37 @@ Status RequestServiceImpl::Send(ServerContext* context, const Request* proto_req
...
   auto callback = [&](cc::SendResult&& send_result) {
     if (not std::holds_alternative<bft::client::Reply>(send_result)) {
-      LOG_INFO(logger_, "Send returned error");
       switch (std::get<uint32_t>(send_result)) {
         case (static_cast<uint32_t>(bftEngine::OperationResult::INVALID_REQUEST)):
+          LOG_INFO(logger_, "Request failed with INVALID_ARGUMENT error for cid=" << req_config.correlation_id);
           status.set_value(grpc::Status(grpc::StatusCode::INVALID_ARGUMENT, "Invalid argument"));
           break;
         case (static_cast<uint32_t>(bftEngine::OperationResult::NOT_READY)):
+          LOG_INFO(logger_, "Request failed with INVALID_ARGUMENT error for cid=" << req_config.correlation_id);
```
Comment (yuliasherman) [inline client/clientservice/src/request_service.cpp:53]: An incorrect error is written to the log. It should be NOT_READY here. Please fix all other logs, too.

Comment (situ-s) [reply client/clientservice/src/request_service.cpp:53]: sure, changes were not added. will add them . thanks

File: tests/simpleKVBC/TesterReplica/internalCommandsHandler.cpp (hunk 13 above)
Comment (situ-s) [reply tests/simpleKVBC/TesterReplica/internalCommandsHandler.cpp:549]: done

File: client/clientservice/src/request_service.cpp (hunk 9 above)
Comment (situ-s) [reply client/clientservice/src/request_service.cpp]: done

File: bftengine/src/preprocessor/PreProcessor.cpp [hunk 15]
```diff
@@ -1164,7 +1164,10 @@ void PreProcessor::finalizePreProcessing(NodeIdType clientId, uint16_t reqOffset
       const auto &span_context = preProcessReqMsg->spanContext<PreProcessRequestMsgSharedPtr::element_type>();
       // Copy of the message body is unavoidable here, as we need to create a new message type which lifetime is
       // controlled by the replica while all PreProcessReply messages get released here.
-      const auto preProcessResult = static_cast<uint32_t>(reqProcessingStatePtr->getAgreedPreProcessResult());
+      auto preProcessResult = static_cast<uint32_t>(reqProcessingStatePtr->getAgreedPreProcessResult());
+      if (preProcessResult == static_cast<uint32_t>(OperationResult::SUCCESS)) {
```
Comment (yuliasherman) [inline bftengine/src/preprocessor/PreProcessor.cpp:1168]: Add a comment on why we do that.

File: tests/simpleKVBC/TesterReplica/internalCommandsHandler.cpp [hunk 16]
```diff
@@ -90,8 +90,7 @@ void InternalCommandsHandler::execute(InternalCommandsHandler::ExecutionRequests
   auto pre_execute = requests.back().flags & bftEngine::PRE_PROCESS_FLAG;
 
   for (auto &req : requests) {
-    if (req.outExecutionStatus != static_cast<uint32_t>(OperationResult::UNKNOWN))
-      continue;  // Request already executed (internal or pre-executed)
+    if (req.outExecutionStatus != static_cast<uint32_t>(OperationResult::UNKNOWN)) continue;
```
Comment (yuliasherman) [inline tests/simpleKVBC/TesterReplica/internalCommandsHandler.cpp:93]: Add a comment: // Request already executed (internal)

File: tests/simpleKVBC/TesterReplica/internalCommandsHandler.cpp [hunk 17]
```diff
@@ -123,7 +122,10 @@ void InternalCommandsHandler::execute(InternalCommandsHandler::ExecutionRequests
                                 merkleUpdates);
     }
     if (res != OperationResult::SUCCESS) LOG_WARN(m_logger, "Command execution failed!");
-    req.outExecutionStatus = static_cast<uint32_t>(res);
+
+    if (req.outExecutionStatus == static_cast<uint32_t>(OperationResult::UNKNOWN)) {
```
Comment (yuliasherman) [inline tests/simpleKVBC/TesterReplica/internalCommandsHandler.cpp:126]: Explain that this way we can pass req.outExecutionStatus set to error by the Apollo test and then verify that error reply gets returned.

Comment (situ-s) [reply tests/simpleKVBC/TesterReplica/internalCommandsHandler.cpp:126]: sure

File: tests/simpleKVBC/TesterReplica/internalCommandsHandler.cpp (hunk 16 above)
Comment (situ-s) [reply tests/simpleKVBC/TesterReplica/internalCommandsHandler.cpp:93]: done

File: bftengine/src/preprocessor/PreProcessor.cpp (hunk 15 above)
Comment (situ-s) [reply bftengine/src/preprocessor/PreProcessor.cpp:1168]: done
